packages/liveblocks-react/src/factory.tsx

```diff
@@ -47,6 +47,49 @@ function getEmptyOthers() {
   return EMPTY_OTHERS;
 }
 
+function makeMutationContext<
+  TPresence extends JsonObject,
+  TStorage extends LsonObject,
+  TUserMeta extends BaseUserMeta,
+  TRoomEvent extends Json
+>(
+  room: Room<TPresence, TStorage, TUserMeta, TRoomEvent>
+): MutationContext<TPresence, TStorage, TUserMeta> {
+  const errmsg =
+    "This mutation cannot be used until connected to the Liveblocks room";
+
+  return {
+    get root() {
+      const root = room.getStorageSnapshot();
+      if (root === null) {
+        throw new Error(errmsg);
+      }
+      return root;
+    },
+
+    get self() {
+      const self = room.getSelf();
+      // NOTE: We could use room.isSelfAware() here to keep the check
+      // consistent with `others`, but we also want to refine the `null` case
+      // away here.
+      if (self === null) {
+        throw new Error(errmsg);
+      }
+      return self;
+    },
```

Comment:
Made these keys lazy, so that if your mutation won't need `self` and doesn't access it, it will not recompute the immutable self value if it would be invalidated at that point.